### ISSUE TITLE
java.io.UncheckedIOException: java.io.EOFException

### ISSUE THREAD

**Issue by HungUnicorn** [dehora/nakadi-java#222]
Our consumer using 0.9.0 stops after seeing this exception. The size of event type is not big.
```
 java.io.UncheckedIOException: java.io.EOFException  
 #011at java.io.BufferedReader$1.hasNext(BufferedReader.java:574)  
 #011at java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1811)  
 #011at java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)  
 #011at nakadi.shadow.io.reactivex.internal.operators.flowable.FlowableFromIterable$IteratorSubscription.fastPath(FlowableFromIterable.java:191)  
 #011at nakadi.shadow.io.reactivex.internal.operators.flowable.FlowableFromIterable$BaseRangeSubscription.request(FlowableFromIterable.java:123)  
 #011at nakadi.shadow.io.reactivex.internal.subscribers.BasicFuseableSubscriber.request(BasicFuseableSubscriber.java:153)  
 #011at nakadi.shadow.io.reactivex.internal.operators.flowable.FlowableOnBackpressureBuffer$BackpressureBufferSubscriber.onSubscribe(FlowableOnBackpressureBuffer.java:91)  
 #011at nakadi.shadow.io.reactivex.internal.subscribers.BasicFuseableSubscriber.onSubscribe(BasicFuseableSubscriber.java:67)  
 #011at nakadi.shadow.io.reactivex.internal.operators.flowable.FlowableFromIterable.subscribe(FlowableFromIterable.java:69)  
 #011at nakadi.shadow.io.reactivex.internal.operators.flowable.FlowableFromIterable.subscribeActual(FlowableFromIterable.java:47)  
 #011at nakadi.shadow.io.reactivex.Flowable.subscribe(Flowable.java:13013)  
 #011at nakadi.shadow.io.reactivex.internal.operators.flowable.FlowableDoOnEach.subscribeActual(FlowableDoOnEach.java:49)  
 #011at nakadi.shadow.io.reactivex.Flowable.subscribe(Flowable.java:13013)  
 #011at nakadi.shadow.io.reactivex.internal.operato448rs.flowable.FlowableOnBackpressureBuffer.subscribeActual(FlowableOnBackpressureBuffer.java:46)  
 #011at nakadi.shadow.io.reactivex.Flowable.subscribe(Flowable.java:13013)  
 #011at nakadi.shadow.io.reactivex.internal.operators.flowable.FlowableMap.subscribeActual(FlowableMap.java:38)  
 #011at nakadi.shadow.io.reactivex.Flowable.subscribe(Flowable.java:13013)  
 #011at nakadi.shadow.io.reactivex.Flowable.subscribe(Flowable.java:12960)  
 #011at nakadi.shadow.io.reactivex.internal.operators.flowable.FlowableUsing.subscribeActual(FlowableUsing.java:73)  
 #011at nakadi.shadow.io.reactivex.Flowable.subscribe(Flowable.java:13013)  
 #011at nakadi.shadow.io.reactivex.Flowable.subscribe(Flowable.java:12960)  
 #011at nakadi.shadow.io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82)  
 #011at nakadi.shadow.io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker$BooleanRunnable.run(ExecutorScheduler.java:260)  
 #011at nakadi.shadow.io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker.run(ExecutorScheduler.java:225)  
 #011at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)  
 #011at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)  
 #011at java.lang.Thread.run(Thread.java:748)  
 Caused by: java.io.EOFException  
 #011at nakadi.shadow.okio.RealBufferedSource.require(RealBufferedSource.java:59)  
 #011at nakadi.shadow.okio.RealBufferedSource.readHexadecimalUnsignedLong(RealBufferedSource.java:293)  
 #011at nakadi.shadow.okhttp3.internal.http1.Http1Codec$ChunkedSource.readChunkSize(Http1Codec.java:444)  
 #011at nakadi.shadow.okhttp3.internal.http1.Http1Codec$ChunkedSource.read(Http1Codec.java:425)  
 #011at nakadi.shadow.okio.RealBufferedSource.read(RealBufferedSource.java:45)  
 #011at nakadi.shadow.okio.RealBufferedSource.exhausted(RealBufferedSource.java:55)  
 #011at nakadi.shadow.okio.InflaterSource.refill(InflaterSource.java:101)  
 #011at nakadi.shadow.okio.InflaterSource.read(InflaterSource.java:62)  
 #011at nakadi.shadow.okio.GzipSource.read(GzipSource.java:80)  
 #011at nakadi.shadow.okio.RealBufferedSource$1.read(RealBufferedSource.java:429)  
 #011at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)  
 #011at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)  
 #011at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)  
 #011at java.io.InputStreamReader.read(InputStreamReader.java:184)  
 #011at nakadi.shadow.okhttp3.ResponseBody$BomAwareReader.read(ResponseBody.java:251)  
 #011at java.io.BufferedReader.fill(BufferedReader.java:161)  
 #011at java.io.BufferedReader.readLine(BufferedReader.java:324)  
 #011at java.io.BufferedReader.readLine(BufferedReader.java:389)  
 #011at java.io.BufferedReader$1.hasNext(BufferedReader.java:571)  
 #011... 26 more  
```

**Comment by dehora** [comment]
> Our consumer using 0.9.0 stops after seeing this exception.

@HungUnicorn can you definitely confirm it stops and doesn't retry? The client's waiting on bytes from the server's response and gives up after a while. I'd expect it to recover and retry, but maybe there's a bug in the 0.9.0 revision.

**Comment by HungUnicorn** [comment]
I found  the retry message two times
 INFO  NakadiClient                                                  - stream_retry_will_sleep sleep=1000 attempt=2/2147483647 thread=nakadi-java-io-0 error=java.io.EOFException  `

but it then encountered the other error
`nakadi.shadow.io.reactivex.exceptions.UndeliverableException: java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@6ffca811 rejected from java.util.concurrent.ScheduledThreadPoolExecutor@3368ca55[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]  `

since then it stopped retrying.

In the old version 0.8.5, I found this java.io.EOFException will call onError(), as we create a new processor, we can keep consuming data. Nevertheless, in 0.9.0 I see onError() is not called when seeing this exception, and therefore we don't create a new processor.

**Comment by satybald** [comment]
I can confirm that we have the same error on 0.9.0 version. It's also related to #198.

**Comment by satybald** [comment]
might it be the case, when Nakadi instance shutdowns, ELB is still having a connection open with the client. Then nakadi-java library is not able to parse properly EOF?

**Comment by satybald** [comment]
we have tons of this exception in our logs.

```
Caused by: java.io.EOFException: \n not found: limit=0 content=…  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:226)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:210)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okhttp3.internal.http1.Http1Codec$ChunkedSource.readChunkSize(Http1Codec.java:441)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okhttp3.internal.http1.Http1Codec$ChunkedSource.read(Http1Codec.java:425)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okio.RealBufferedSource.read(RealBufferedSource.java:45)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okio.RealBufferedSource.exhausted(RealBufferedSource.java:55)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okio.InflaterSource.refill(InflaterSource.java:101)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okio.InflaterSource.read(InflaterSource.java:62)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okio.GzipSource.read(GzipSource.java:80)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okio.RealBufferedSource$1.read(RealBufferedSource.java:429)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at java.io.InputStreamReader.read(InputStreamReader.java:184)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at nakadi.shadow.okhttp3.ResponseBody$BomAwareReader.read(ResponseBody.java:251)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at java.io.BufferedReader.fill(BufferedReader.java:161)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at java.io.BufferedReader.readLine(BufferedReader.java:324)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at java.io.BufferedReader.readLine(BufferedReader.java:389)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011at java.io.BufferedReader$1.hasNext(BufferedReader.java:571)  
12:13:35.060 diga-dev /var/log/application.log  Sep  4 10:13:34 ip-172-31-131-100 docker/9c7cbe07483f[827]: #011... 26 common frames omitted  

```

**Comment by dehora** [comment]
The stop behaviour should be addressed since 0.9.6. When the processor is stopped there can be errors thrown around connections and threadpools due to the way interrupts are occurring when a new processor is started but they themselves don't stop the new processor from starting; 0.9.9 tries  to reduce the amount of noise they create. Closing for now.
